llvm/lib/Transforms/Vectorize/VPlanValue.h

```diff
@@ -60,9 +63,13 @@ class LLVM_ABI_FOR_TEST VPValue {
   /// Hold the underlying Value, if any, attached to this VPValue.
   Value *UnderlyingVal;
 
-  /// Pointer to the VPDef that defines this VPValue. If it is nullptr, the
-  /// VPValue is not defined by any recipe modeled in VPlan.
-  VPDef *Def;
+  /// Pointer to the VPDef that defines this VPValue, or VPRegionBlock for
+  /// VPRegionValue. If it is nullptr, the VPValue is not defined by any recipe
+  /// or region in VPlan, i.e. is a live-in.
+  union {
+    VPDef *Def;
+    VPRegionBlock *DefiningRegion;
```

Comment:
If DefiningRegion is held in the VPValue base class, so should `getDefiningRegion()`, complementing `getDefiningRecipe()`?

If `hasDefiningRegion()` is also supported, complementing `hasDefiningRecipe()`, is it worth having a VPRegionValue as a derived class? OTOH, would be consistent to also have a VPLiveIn as another derived class of VPValue?

Reply:
Would be good for the new region-defined VPValues to have an API that is consistent with the API's of live-ins and recipe-defined VPValues. Aligning the implementation to hold UnderlyingValue inside the union seems like a good cleanup either before or after this patch.